[CORE-13986] `storage`: expose `storage_log_segment_size` histogram internal metric

WillemKauf opened the pull request:
This PR exposes a `storage_log_segment_size` histogram which is aggregated across shards/partitions to give visibility into the profile of `segment` sizes for a given topic. 

This histogram is calculated by iterating over all `segment`s in a partition every `storage_segment_size_refresh_rate_ms` and recording their sizes in a `hdr_hist`. This interval defaults to `60min`, since the profile of segment sizes is not expected to fluctuate wildly, nor often in most cases. The `O(n)` computation of the histogram can be disabled by setting this cluster config to `std::nullopt`.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* Expose the `vectorized_storage_log_segment_size` internal metric for visibility into the distribution of `segment` sizes within a topic.

WillemKauf committed:
`ssx`: support different `Rate` types in `rate_limited_function`

The inspiration here is to be able to support more complex types than
just a bare duration type within the `rate_limited_function`. For example,
for a duration type `T`, a motivating case would be a `binding<T>` to a
cluster property controlling an interval rate:

```
auto fn = ssx::rate_limited_function<
   void,
   seastar::lowres_clock,
   config::binding<std::chrono::milliseconds>>(
   [this] {
       foo();
   },
   config::shard_local_cfg().foo_interval.bind());
```

or an `optional`:

```
auto fn = ssx::rate_limited_function<
   void,
   seastar::lowres_clock,
   std::optional<std::chrono::milliseconds>>(
   [this] {
       foo();
   },
   config::shard_local_cfg().optional_foo_interval()),
```

or a `binding<std::optional<T>>`, or even something like a
`ss::noncopyable_function<T()`>, and so on and so forth.

The changes in this commit allow for these more complex types to be used
within the `rate_limited_function`.

WillemKauf commented on src/v/ssx/rate_limited_function.h:
```diff
@@ -47,12 +51,27 @@ class rate_limited_function<Ret() noexcept(Noexcept), Clock> {
     // Returns true if the cached value is ready to be used.
     bool is_ready(Clock::time_point now = Clock::now()) const {
-        return _result.has_value() && now <= (_last_call + _rate);
+        auto rate = [](
+                      this auto&& self,
+                      auto&& rate) -> std::optional<typename Clock::duration> {
+            if constexpr (std::is_invocable_v<decltype(rate)>) {
+                return self(rate());
+            } else {
+                return rate;
+            }
+        }(_rate);
```
but i like recursive lambdas

WillemKauf commented:
Lots of `pyright` strict mode failures for newly added file, I'll try to fix up the type checking tomorrow. Rest of PR should be reviewed as is.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74233-->
<!--7ae50ad0-1399-48b5-81cc-2760cdc6a4db-->
<!--42ce8b54-7877-4b66-bb9e-1b2c871a7625-->
<!--1cfc26f5-bca2-4157-9ada-31c72fc6ab55-->
<!--9538fad3-581f-4664-85af-9dd11fde980d-->
<!--title-->
## Retry command for [Build#74233](https://buildkite.com/redpanda/redpanda/builds/74233)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/controller_partition_invariants_test.py::ControllerLogInvariantsTest.test_controller_is_not_compacted_nor_deleted@{"cluster_cleanup_policy":"delete"}
tests/rptest/tests/controller_partition_invariants_test.py::ControllerLogInvariantsTest.test_controller_is_not_compacted_nor_deleted@{"cluster_cleanup_policy":"compact"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4bbd4ed5-8193-4576-a572-b245948218d5-->
<!--TEST RESULTS-->
<!--5016bbe4-0ae9-4d8e-b476-c26042aebbbd-->
<!--TEST RESULTS-->
<!--602ab477-36cb-46af-9408-0c521620bd5a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74233">build#74233</a></summary>


| test_class                  | test_method                                  | test_arguments                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                           |
|:----------------------------|:---------------------------------------------|:--------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogInvariantsTest | test_controller_is_not_compacted_nor_deleted | {"cluster_cleanup_policy": "compact"} | integration | https://buildkite.com/redpanda/redpanda/builds/74233#0199e673-ef11-4058-b637-875a300f98c9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogInvariantsTest&test_method=test_controller_is_not_compacted_nor_deleted |
| ControllerLogInvariantsTest | test_controller_is_not_compacted_nor_deleted | {"cluster_cleanup_policy": "compact"} | integration | https://buildkite.com/redpanda/redpanda/builds/74233#0199e67f-6dee-4baa-ae16-5a6da07e7e94 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogInvariantsTest&test_method=test_controller_is_not_compacted_nor_deleted |
| ControllerLogInvariantsTest | test_controller_is_not_compacted_nor_deleted | {"cluster_cleanup_policy": "delete"}  | integration | https://buildkite.com/redpanda/redpanda/builds/74233#0199e673-ef12-4807-8c0d-94a312f3a8b0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogInvariantsTest&test_method=test_controller_is_not_compacted_nor_deleted |
| ControllerLogInvariantsTest | test_controller_is_not_compacted_nor_deleted | {"cluster_cleanup_policy": "delete"}  | integration | https://buildkite.com/redpanda/redpanda/builds/74233#0199e67f-6def-47b9-9806-716dba296d5c | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogInvariantsTest&test_method=test_controller_is_not_compacted_nor_deleted |
| TimeQueryTest               | test_timequery_with_local_gc                 | null                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74233#0199e67f-6df4-4d8a-8ca4-a77b7bbdd644 | FLAKY         | 19/21    | upstream reliability is '90.15817223198594'. current run reliability is '90.47619047619048'. drift is -0.31802 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74329">build#74329</a></summary>


| test_class                      | test_method                          | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                       | test_history                                                                                                                                                             |
|:--------------------------------|:-------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupTest               | test_dead_group_recovery             | {"static_members": false}                          | integration | https://buildkite.com/redpanda/redpanda/builds/74329#0199ee5d-3426-4315-8bb8-df4fd93b4187 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupTest&test_method=test_dead_group_recovery                 |
| ConsumerGroupTest               | test_dead_group_recovery             | {"static_members": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74329#0199ee5d-3427-452a-b6ea-544eeb8403f9 | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupTest&test_method=test_dead_group_recovery                 |
| NodesDecommissioningTest        | test_recycle_all_nodes               | {"auto_assign_node_id": false}                     | integration | https://buildkite.com/redpanda/redpanda/builds/74329#0199ee5d-3426-4315-8bb8-df4fd93b4187 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recycle_all_nodes            |
| PartitionBalancerTest           | test_partition_autobalancer_sanction | {"disable_license": true}                          | integration | https://buildkite.com/redpanda/redpanda/builds/74329#0199ee5d-3428-498b-8700-e815932ef32f | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_partition_autobalancer_sanction |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test             | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74329#0199ee5d-3433-4d09-8c30-f2f1a231ff14 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test             | {"cloud_storage_type": 1, "mixed_versions": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74329#0199ee5d-3426-4315-8bb8-df4fd93b4187 | FLAKY         | 13/21    | upstream reliability is '100.0'. current run reliability is '61.904761904761905'. drift is 38.09524 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74367">build#74367</a></summary>


| test_class                      | test_method              | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74367#0199f049-c855-4258-9a0e-2fbf402155a4 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<!--/body-->

StephanDollberg reviewed (commented):
I don't think per partition/topic histograms are feasible. This would bump the total metrics count for non-aggregate metric clusters by 30% or something.

Even on a topic only level it would be a problem for large topic count clusters. (cc @ballard26)

Can we do:
 - Add a global histogram metric (preaggregated outside the metrics layer) 
 - Make the per topic info be available via the admin API

?

Lazin commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -4623,4 +4626,18 @@ bool disk_log_impl::needs_compaction() const {
+    }
+
+    for (const auto& seg : _segs) {
+        hist.record(seg->size_bytes());
```
I'm not entirely convinced that it's a good idea to recompute the histogram. Could be expensive if we have very large number of partitions with few segments each.

WillemKauf replied:
> I'm not entirely convinced that it's a good idea to recompute the histogram

Unfortunately recomputation feels necessary. This is why I have the tweakable cluster config defaulting at a relatively infrequent interval, and which can also completely disable the calculation.

Lazin commented on src/v/storage/probe.cc:
```diff
@@ -248,6 +251,19 @@ void probe::setup_metrics(const model::ntp& ntp) {
+            config::shard_local_cfg()
+              .storage_segment_size_refresh_rate_ms.bind()),
+          sm::description("Local segment size histogram in bytes."),
+          labels),
```
Can we do this not per ntp but per broker or at least per shard? A histogram per ntp could be quite expensive in terms of metrics cardinality.

WillemKauf replied:
Yes, we could aggregate across all shards/hosted partitions by moving this calculation to the `log_manager`.

WillemKauf commented:
>I don't think per partition/topic histograms are feasible. This would bump the total metrics count for non-aggregate metric clusters by 30% or something.

Yeah I suspected as much, thanks for confirming

> Add a global histogram metric (preaggregated outside the metrics layer)

We could perform the calculation in the `log_manager` and aggregate across all hosted partitions & shards on this broker.

dotnwat commented:
> > I don't think per partition/topic histograms are feasible. This would bump the total metrics count for non-aggregate metric clusters by 30% or something.
> 
> Yeah I suspected as much, thanks for confirming
> 
> > Add a global histogram metric (preaggregated outside the metrics layer)
> 
> We could perform the calculation in the `log_manager` and aggregate across all hosted partitions & shards on this broker.

Can we take advantage of the work that space management already does related to stat-ing and caching the size of segments, or perhaps something really simple like updating the histogram when we roll a segment?

@StephanDollberg per-shard would be ok?

WillemKauf commented:
> something really simple like updating the histogram when we roll a segment?

This wouldn't take into account compaction of older segments, or prefix truncation/removal of old segments, etc.

Recomputing the histogram over all segments seems like the only way to get an accurate reading, though you are right, we may be able to optimize by piggy-backing off of space management calculations or something like that.

dotnwat commented:
> Recomputing the histogram over all segments seems like the only way to get an accurate reading, though you are right, we may be able to optimize by piggy-backing off of space management calculations or something like that.

Recomputing the entire histogram is one option. But wouldn't the other option be to update the histogram when a new segment is created with a stable size (rolling, compaction, etc...) and ignore cases like removal of segments? I say that because the histogram state is effectively rolling, rather than a point-in-time accurate snapshot?

WillemKauf commented:
>stable size... ignore cases like removal of segments

Take for example a compacted topic with adjacent segment merging. At first, the histogram reflects say 5 segments each `128_MiB` in size. 

Then, they are compacted down such that the first 4 segments are `1_MiB` each and the last segment is now `100_MiB`. Updating the histogram by adding the newly compacted segments now shows there are 10 segments total, with most segments being `128_MiB` in size (which is clearly not reflective of reality). 

Then, imagine we merge the first 4 segments to create a new segment which is `4_MiB` in size and add that to the histogram. Now the histogram shows 11 segments total with most segments still being `128_MiB` in size and 4 small segments being `1_MiB` in size when in reality, there are two segments which are `4_MiB` and `100_MiB` large.

Over time this problem will compound and give an entirely inaccurate portrayal of the segment size distribution due to the fact segment sizes are not static/stable at all when closed in either a `compact` or `delete` topic.

StephanDollberg commented:
> per-shard would be ok?

Yeah for sure.

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -1620,6 +1620,14 @@ configuration::configuration()
+      "storage_segment_size_refresh_rate_ms",
+      "The rate at which the local storage histogram for the internal metric "
+      "`segment_size` is recomputed. If set to `std::nullopt`, recomputation "
+      "of this metric is disabled.",
```
```suggestion
      "The interval, in milliseconds, for recalculating the"
      " `storage_log_segment_size` histogram metric. This metric provides"
      " visibility into segment size distribution across topics. "
      "Set to `null` to disable histogram recalculation.",
```

Changed to active voice. Used "recalculation" instead of "recomputation" as it's not very common to see this used. Also removed the c++ nullopt to "null". 

Feel free to accept what makes more sense to you.

WillemKauf committed:
`config`: add `storage_segment_size_refresh_rate_ms`

The rate at which the local storage histogram for the internal metric
`segment_size` is recomputed from within the `log_manager`. If set to
`std::nullopt`, recomputation of this metric is disabled.

This is a potentially expensive calculation, which is why the default
value is relatively infrequent.

WillemKauf committed:
`storage`: add `segment_size_histogram` related members to `log_manager`

Adds `schedule_segment_size_histogram()`, `try_get_segment_size_histogram()`,
and `compute_segment_size_histogram()` member functions, and a member variable
`_segment_size_hist_fut`.

The idea here is that `compute_segment_size_histogram()` is potentially
an expensive loop to perform synchronously, and is therefore made
asynchronous to prevent reactor stalls. Therefore, when a new segment size
histogram is requested from e.g. a storage layer `probe`, a backgrounded
future for computing this histogram is launched and later resolved.

WillemKauf committed:
`storage`: add `segment_size` metric to `log_manager_probe`

This exposes a histogram of segment sizes on a given shard to metrics.

`log_manager::schedule_calc_segment_size_histogram()` is passed into an
`ssx::rate_limited_function` on the interval `storage_segment_size_refresh_rate_ms`,
while the cached `_segment_size_histogram` is held in memory.

WillemKauf committed:
`storage`: add `log_manager_segment_size_histogram` fixture tests

WillemKauf committed:
`rptest`: add `log_segment_size_test`

WillemKauf commented:
Reworked this PR by pushing the calculation to the `log_manager` and also aggregating by shard when `aggregate_metrics=true`.

This solution uses some funky asynchronous scheduling to compute a new `segment_size_histogram` in the background on the interval `storage_segment_size_refresh_rate_ms`.

The function `log_manager::compute_segment_size_histogram()` is asynchronous in order to allow yielding to the reactor- this does complicate the implementation, as can be seen in the changes.

I would propose that if we _don't_ want this complicated a solution, that we would have to investigate statistical sampling of a sub-set of `segment`s per `log` instead of iteration over _all_ `segment`s in _all_ `log`s per shard, in order to go back to a synchronous calculation, as noted by a `TODO` comment:

> // TODO: Uniform random sampling of segments to reduce cost? Could
> // potentially avoid any asychronicity if the calculation had a clear upper
> // bound.

Let me know what you think reviewers, I would understand if the complexity here is undesirable for what should be a fairly simple task.

WillemKauf commented:
FWIW, i can also see a world in which this metric is mostly useless and we end up paying for something we don't use- in which case an admin endpoint to request the segment size histogram for a given partition is probably more useful.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74329-->
<!--97234922-c025-4837-b515-147244984baa-->
<!--76916307-7dcc-404b-b533-7616540fbaab-->
<!--title-->
## Retry command for [Build#74329](https://buildkite.com/redpanda/redpanda/builds/74329)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_dead_group_recovery@{"static_members":true}
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_dead_group_recovery@{"static_members":false}
```
<!--/body-->
